### PR TITLE
Add `maketime` and `makedate`

### PR DESCRIPTION
**Pull request by Yury-Fridlyand**
### Description
* `maketime` [link](https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html#function_maketime) [link](https://www.techonthenet.com/mysql/functions/maketime.php)
  * Supports any values in range `00:00:00.0 - 23:59:59.(9)`;
  * All arguments are double, but hour and minute are rounded (as MySQL works);
  * Fraction second part is kept;
  * NULL values causes NULL result;
  * Unlike MySQL, negative values, and hour > 23 are not supported. Implementation of that requires big changes.
* `makedate` [link](https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html#function_makedate) [link](https://www.techonthenet.com/mysql/functions/makedate.php)
  * All arguments are double, but rounded (as MySQL works);
  * Zero year interpreted as 2000 (as MySQL works);
  * NULL values causes NULL result;
  * day < 1 causes NULL result;
  * day > 365/366 causes switching to the next year.

Functions have 2 implementations:
* easier
* faster

I used `easier` one, but `faster` one used in unit tests. Could be swapped.

According to MySQL both functions accept double values. Arguments which are not supposed to have fraction parts (e.g. hour, year) are rounded.

### TODO
- [x] Implementation
- [x] Unit tests
- [x] Integration tests
- [x] Doctests
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass, including unit test, integration test and doctest
- [x] New functionality has been documented.
  - [ ] New functionality has javadoc added
  - [x] New functionality has user manual doc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java [hunk 1]
```diff
@@ -512,6 +529,28 @@ private ExprValue exprHour(ExprValue time) {
...
+      return ExprNullValue.of();
+    }
+    if (0 == year) {
+      year = 2000;
+    }
+    return new ExprDateValue(LocalDate.ofYearDay((int)year, 1).plusDays(dayOfYear - 1));
+  }
+
+  private ExprValue exprMakeTime(ExprValue hour, ExprValue minute, ExprValue second) {
+    if (0 > hour.doubleValue() || 0 > minute.doubleValue() || 0 > second.doubleValue()) {
```
Comment (MitchellGale-BitQuill) [inline core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java:546]: Can the minutes and seconds not be valid at 0?

**Review by MitchellGale-BitQuill — APPROVED** [review]

File: core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java (hunk 1 above)
Comment (Yury-Fridlyand) [reply core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java:546]: I check here for negative values, zero is OK for `maketime`:
```
opensearchsql> select maketime(0, 0, 0);
fetched rows / total rows = 1/1
+---------------------+
| maketime(0, 0, 0)   |
|---------------------|
| 00:00:00            |
+---------------------+
```
reference:
```
mysql> select maketime(0, 0, 0);
+-------------------+
| maketime(0, 0, 0) |
+-------------------+
| 00:00:00          |
+-------------------+
1 row in set (0.00 sec)
```

File: core/src/test/java/org/opensearch/sql/expression/datetime/DateTimeFunctionTest.java [hunk 2]
```diff
@@ -994,6 +1006,139 @@ public void date_format() {
...
+    return maketime(DSL.literal(hour), DSL.literal(minute), DSL.literal(second))
+        .valueOf(null).timeValue();
+  }
+
+  @Test
+  public void maketime() {
+    var r = new Random();
+    assertEquals(maketime(20., 30., 40.), LocalTime.of(20, 30, 40));
+
+    for (var ignored : IntStream.range(0, 25).toArray()) {
```
Comment (MaxKsyunz) [inline core/src/test/java/org/opensearch/sql/expression/datetime/DateTimeFunctionTest.java:1026]: Why is this for loop necessary? What does it test that isn't covered by the test on line 1024 above?

As an aside, tests must be reproducible so if a test uses RNG, the user must be able to seed it and any test failures must include the seed used. Otherwise, if it does fail, there is no way to investigate the failure.

File: core/src/test/java/org/opensearch/sql/expression/datetime/DateTimeFunctionTest.java [hunk 3]
```diff
@@ -994,6 +1006,139 @@ public void date_format() {
...
+
+  /**
+   * Combines permutations of @size args with @exprToPermute placed in all possible position(s),
+   * the rest are filled with @exprToFill.
+   * @param size Size of the array to permute
+   * @param exprToPermute Expression to permute, usually invalid one to test
+   * @param exprToFill Expression to fill the rest place, usually valid one
+   * @return A collection with all possible permutations
+   */
+  private Set<Expression[]> permuteArgumentsWith(int size,
```
Comment (MaxKsyunz) [inline core/src/test/java/org/opensearch/sql/expression/datetime/DateTimeFunctionTest.java:1125]: The tests would be easier to understand if the permutations tested were listed. 

Since this function is only used with `size` of 2 and 3, I'd replace it with a function that just uses `List.of(...)` and lists all the permutations.

File: core/src/test/java/org/opensearch/sql/expression/datetime/DateTimeFunctionTest.java [hunk 4]
```diff
@@ -994,6 +1006,139 @@ public void date_format() {
...
+          String.format("year = %f, dayOfYear = %f", year, dayOfYear));
+    }
+
+    assertEquals(LocalDate.ofYearDay(2000, 42), makedate(0., 42.),
+        "0 year is not interpreted as 2000 as in MySQL");
+    assertEquals(nullValue(), eval(makedate(DSL.literal(-1.), DSL.literal(42.))),
+        "Negative year doesn't produce NULL");
+    assertEquals(nullValue(), eval(makedate(DSL.literal(42.), DSL.literal(-1.))),
+        "Negative dayOfYear doesn't produce NULL");
+    assertEquals(nullValue(), eval(makedate(DSL.literal(42.), DSL.literal(0.))),
```
Comment (MaxKsyunz) [inline core/src/test/java/org/opensearch/sql/expression/datetime/DateTimeFunctionTest.java:1103]: What about day 366 of a non-leap year?

File: core/src/test/java/org/opensearch/sql/expression/datetime/DateTimeFunctionTest.java [hunk 5]
```diff
@@ -994,6 +1006,139 @@ public void date_format() {
...
+  }
+
+  private LocalDate makedate(Double year, Double dayOfYear) {
+    return makedate(DSL.literal(year), DSL.literal(dayOfYear)).valueOf(null).dateValue();
+  }
+
+  @Test
+  public void makedate() {
+    var r = new Random();
+    for (var ignored : IntStream.range(0, 1025).toArray()) {
```
Comment (MaxKsyunz) [inline core/src/test/java/org/opensearch/sql/expression/datetime/DateTimeFunctionTest.java:1071]: What cases are tested in this for loop? Why is RNG necessary?

File: docs/user/dql/functions.rst [hunk 6]
```diff
@@ -1219,15 +1219,60 @@ Example::
...
 
 
+MAKEDATE
+--------
+
+Description
+>>>>>>>>>>>
+
+Returns a date, given `year` and `day-of-year` values. `dayofyear` must be greater than 0 or the result is `NULL`. The result is also `NULL` if either argument is `NULL`.
+Arguments are rounded to an integer.
```
Comment (MaxKsyunz) [inline docs/user/dql/functions.rst:1229]: Was there a test case for this rounding behaviour in DateTimeFunctionTest?

File: docs/user/dql/functions.rst [hunk 7]
```diff
@@ -1219,15 +1219,60 @@ Example::
...
+Return type: DATE
+
+Example::
+
+    os> select MAKEDATE(1945, 5.9) as f1, MAKEDATE(1984, 1984) as f2
+    fetched rows / total rows = 1/1
+    +------------+------------+
+    | f1         | f2         |
+    |------------+------------|
+    | 1945-01-06 | 1989-06-06 |
```
Comment (MaxKsyunz) [inline docs/user/dql/functions.rst:1246]: If sample code is going to show that `MAKEDATE(1984, 1984) is 1989-...` the description should describe this behaviour as well.

This is also a good test case to have in the unit tests -- that `MAKEDATE($year, $days_in_year + 1) == MAKEDATE($year + 1, 1)`.

File: docs/user/dql/functions.rst [hunk 8]
```diff
@@ -1219,15 +1219,60 @@ Example::
...
+
+
 MAKETIME
 --------
 
 Description
 >>>>>>>>>>>
 
+Returns a time value calculated from the hour, minute, and second arguments. Returns `NULL` if any of its arguments are `NULL`.
+The second argument can have a fractional part, rest arguments are rounded to an integer.
```
Comment (MaxKsyunz) [inline docs/user/dql/functions.rst:1257]: Is the fractional part rounded to a particular number of decimal places?

File: docs/user/ppl/functions/datetime.rst [hunk 9]
```diff
@@ -388,15 +388,60 @@ Example::
     +--------------------------+
 
 
+MAKEDATE
```
Comment (MaxKsyunz) [inline docs/user/ppl/functions/datetime.rst:391]: Same comments as in SQL apply here.

File: core/src/test/java/org/opensearch/sql/expression/datetime/DateTimeFunctionTest.java [hunk 10]
```diff
@@ -994,6 +1006,139 @@ public void date_format() {
     assertEquals(missingValue(), eval(dsl.date_format(DSL.literal(""), missingRef)));
   }
 
+  private FunctionExpression maketime(Expression hour, Expression minute, Expression second) {
```
Comment (MaxKsyunz) [inline core/src/test/java/org/opensearch/sql/expression/datetime/DateTimeFunctionTest.java:1009]: I realize these tests follow the convention in this file, but the tests would be much easier to understand if each function had it's own test suite, and each test in a suite tested a particular scenario -- all valid arguments, null as an argument, negative values, rounding behaviour, etc.

I'll leave it up to you whether you want to break with the convention and introduce a MakeTimeTests.

**Review by MaxKsyunz — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java [hunk 11]
```diff
@@ -236,6 +243,16 @@ private FunctionResolver hour() {
     );
   }
 
+  private FunctionResolver makedate() {
+    return define(BuiltinFunctionName.MAKEDATE.getName(),
+        impl(nullMissingHandling(DateTimeFunction::exprMakeDate), DATE, DOUBLE, DOUBLE));
```
Comment (acarbonetto) [inline core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java:248]: Why DOUBLE?  These should be integers, no?  I don't know what makedate(1.5, 1.5) means.

File: core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java [hunk 12]
```diff
@@ -236,6 +243,16 @@ private FunctionResolver hour() {
...
+  }
+
+  private FunctionResolver maketime() {
+    return define(BuiltinFunctionName.MAKETIME.getName(),
+        impl(nullMissingHandling(DateTimeFunction::exprMakeTime), TIME, DOUBLE, DOUBLE, DOUBLE));
```
Comment (acarbonetto) [inline core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java:253]: Why DOUBLE (for hour/minutes), I don't know what maketime(1.5, 1.5, 1.3456) means (at least for hour and minutes)

File: core/src/test/java/org/opensearch/sql/expression/datetime/DateTimeFunctionTest.java (hunk 10 above)
Comment (acarbonetto) [reply core/src/test/java/org/opensearch/sql/expression/datetime/DateTimeFunctionTest.java:1009]: This is a good suggestion.  I would recommend setting the convention for the test file to include the original name, such as: `DateTimeFunctionMakedateTest` and `DateTimeFunctionMaketimeTest`.  It would be easier to find for future developers.

File: core/src/test/java/org/opensearch/sql/expression/datetime/DateTimeFunctionTest.java (hunk 2 above)
Comment (acarbonetto) [reply core/src/test/java/org/opensearch/sql/expression/datetime/DateTimeFunctionTest.java:1026]: Instead of testing RNG numbers, test boundary cases.  For example: 0, and 23:59:59.99 would be decent test cases for time. 
Feb 29 on leap years, and older years would be good boundary cases for dates. 
Lots of timezones (and over the the dateline) are also excellent test cases when timezones are included.

File: docs/user/dql/functions.rst (hunk 6 above)
Comment (acarbonetto) [reply docs/user/dql/functions.rst:1229]: What about invalid positive numbers?  such as dayofyear `999`?

File: core/src/test/java/org/opensearch/sql/expression/datetime/DateTimeFunctionTest.java [hunk 13]
```diff
@@ -994,6 +1006,139 @@ public void date_format() {
...
+    }
+
+    assertEquals(LocalDate.ofYearDay(2000, 42), makedate(0., 42.),
+        "0 year is not interpreted as 2000 as in MySQL");
+    assertEquals(nullValue(), eval(makedate(DSL.literal(-1.), DSL.literal(42.))),
+        "Negative year doesn't produce NULL");
+    assertEquals(nullValue(), eval(makedate(DSL.literal(42.), DSL.literal(-1.))),
+        "Negative dayOfYear doesn't produce NULL");
+    assertEquals(nullValue(), eval(makedate(DSL.literal(42.), DSL.literal(0.))),
+        "Zero dayOfYear doesn't produce NULL");
```
Comment (acarbonetto) [inline core/src/test/java/org/opensearch/sql/expression/datetime/DateTimeFunctionTest.java:1104]: what about tests for invalid hour/minute/second or year/month/day? 
(is there an invalid year, or is year 0 and 9999 valid?)

File: core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java (hunk 11 above)
Comment (acarbonetto) [reply core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java:248]: These are rounded...?

File: core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java (hunk 12 above)
Comment (acarbonetto) [reply core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java:253]: These are rounded?

File: docs/user/dql/functions.rst (hunk 8 above)
Comment (acarbonetto) [reply docs/user/dql/functions.rst:1257]: What about hour > 24?  What about minute/second > 60? 
Make sure to indicate 24 hour time is used, not 12-hour with am/pm.

**Review by acarbonetto — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java (hunk 11 above)
Comment (Yury-Fridlyand) [reply core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java:248]: MySQL accepts doubles and round them, I did the same. Yes, I was surprised too.

File: core/src/test/java/org/opensearch/sql/expression/datetime/DateTimeFunctionTest.java (hunk 13 above)
Comment (Yury-Fridlyand) [reply core/src/test/java/org/opensearch/sql/expression/datetime/DateTimeFunctionTest.java:1104]: MySQL is limited on output value: from 0001-01-01 to 9999-12-31. Java has no limits.

File: core/src/test/java/org/opensearch/sql/expression/datetime/DateTimeFunctionTest.java (hunk 2 above)
Comment (Yury-Fridlyand) [reply core/src/test/java/org/opensearch/sql/expression/datetime/DateTimeFunctionTest.java:1026]: I dump values used on case of failure, but if it not enough I can hardcode values gotten from RNG.

File: core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java (hunk 11 above)
Comment (acarbonetto) [reply core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java:248]: That's... unexpected.  
I need to learn to expect the unexpected from SQL language spec.

File: core/src/test/java/org/opensearch/sql/expression/datetime/DateTimeFunctionTest.java (hunk 5 above)
Comment (Yury-Fridlyand) [reply core/src/test/java/org/opensearch/sql/expression/datetime/DateTimeFunctionTest.java:1071]: RNG helps to find unconsidered cases, it helped me to find few bugs. I dump values which cause a failure, so any other developer would be able to reproduce it.

File: docs/user/dql/functions.rst (hunk 6 above)
Comment (Yury-Fridlyand) [reply docs/user/dql/functions.rst:1229]: > Was there a test case for this rounding behaviour in DateTimeFunctionTest?

Thanks, I will add.

> What about invalid positive numbers? such as dayofyear `999`?

There is no limit! Freedom!

File: core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java [hunk 14]
```diff
@@ -512,6 +528,28 @@ private ExprValue exprHour(ExprValue time) {
     return new ExprIntegerValue(time.timeValue().getHour());
   }
 
+  private ExprValue exprMakeDate(ExprValue yearExpr, ExprValue dayOfYearExp) {
```
Comment (forestmvey) [inline core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java:531]: A javadoc with some explanation of the mysql compliance would be nice.

**Review by forestmvey — APPROVED** [review]

File: core/src/test/java/org/opensearch/sql/expression/datetime/DateTimeFunctionTest.java (hunk 5 above)
Comment (acarbonetto) [reply core/src/test/java/org/opensearch/sql/expression/datetime/DateTimeFunctionTest.java]: A good discussion on using RNG in unit/integration tests: https://softwareengineering.stackexchange.com/questions/429601/is-the-usage-of-random-values-in-unit-testing-a-good-practice

File: core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java [hunk 15]
```diff
@@ -512,6 +528,50 @@ private ExprValue exprHour(ExprValue time) {
...
+   *
+   * @param yearExpr year
+   * @param dayOfYearExp day of the @year, starting from 1
+   * @return Date - ExprDateValue object with LocalDate
+   */
+  private ExprValue exprMakeDate(ExprValue yearExpr, ExprValue dayOfYearExp) {
+    var year = Math.round(yearExpr.doubleValue());
+    var dayOfYear = Math.round(dayOfYearExp.doubleValue());
+    // We need to do this to comply with MySQL
+    if (0 >= dayOfYear || 0 > year) {
```
Comment (acarbonetto) [inline core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java:547]: nit: Can we make this check prior to rounding?  Then if the value is negative, we save those operations.

**Review by acarbonetto — APPROVED** [review]

File: core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java (hunk 15 above)
Comment (Yury-Fridlyand) [reply core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java:547]: To optimize performance we shouldn't use `LocalDate::parse` nor `LocalDate::plusDays`. See my note about faster and easier implementation.

**Review by MaxKsyunz — APPROVED** [review]
